Frameworks/AddressBook/ABRecord.mm

```diff
@@ -80,18 +150,458 @@ ABRecordType ABRecordGetRecordType(ABRecordRef record) {
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Records can only be modified if they are newly created (with ABPersonCreate)
+ or are retrieved with ABAddressBookCopyArrayOfAllUserAppPeople -- contacts from
+ ABAddressBookCopyArrayOfAllPeople are read-only and cannot have their properties changed.
 */
-bool ABRecordSetValue(ABRecordRef record, ABPropertyID property, CFTypeRef value, CFErrorRef* error) {
-    UNIMPLEMENTED();
-    return StubReturn();
+bool ABRecordSetValue(ABRecordRef record, ABPropertyID contactProperty, CFTypeRef value, CFErrorRef* error) {
+    if (record == nullptr) {
+        if (error) {
+            NSDictionary* userInfo = @{
+                NSLocalizedDescriptionKey : NSLocalizedString(@"Error setting record value.\n", nil),
+                NSLocalizedFailureReasonErrorKey : NSLocalizedString(@"The record should not be null!\n", nil)
+            };
+            *error = (__bridge_retained CFErrorRef)[NSError errorWithDomain:(__bridge NSString*)ABAddressBookErrorDomain
+                                                                       code:kABOperationNotPermittedByStoreError
+                                                                   userInfo:userInfo];
+        }
+
+        return false;
+    }
+
+    _ABContact* person = (__bridge _ABContact*)record;
+
+    if (person.type == kAddressBookReadOnlyContact) {
+        if (error) {
+            NSDictionary* userInfo = @{
+                NSLocalizedDescriptionKey : NSLocalizedString(@"Error setting record value.\n", nil),
+                NSLocalizedFailureReasonErrorKey : NSLocalizedString(@"You can only modify contacts created by your app!\n", nil)
+            };
+            *error = (__bridge_retained CFErrorRef)[NSError errorWithDomain:(__bridge NSString*)ABAddressBookErrorDomain
+                                                                       code:kABOperationNotPermittedByStoreError
+                                                                   userInfo:userInfo];
+        }
+
+        return false;
+    }
+
+    // Cases for various name properties.
+    if (contactProperty == kABPersonFirstNameProperty) {
```

Comment:
Since they're exported constant symbols, not defines or enums, they cannot be case labels :(

---
In reply to: [76669898](https://github.com/Microsoft/WinObjC/pull/935#discussion_r76669898) [](ancestors = 76669898)